### DISCUSSION TITLE
Minor pen vector in maintenance mode handling (1.4.x -> 2.2x)

### DISCUSSION THREAD

**Discussion by adamcameron** [wheels-dev/wheels#1269] [General]
Howdy
Have a look at this code: https://github.com/cfwheels/cfwheels/blob/develop/wheels/events/onrequeststart.cfm#L74-L94

```
	if (application.wheels.environment == "maintenance") {
		if (StructKeyExists(url, "except")) {
			application.wheels.ipExceptions = url.except;
		}
		local.makeException = false;
		if (Len(application.wheels.ipExceptions)) {
			if (ReFindNoCase("[a-z]", application.wheels.ipExceptions)) {
				if (ListFindNoCase(application.wheels.ipExceptions, cgi.http_user_agent)) {
					local.makeException = true;
				}
			} else {
				local.ipAddress = cgi.remote_addr;
				if (Len(request.cgi.http_x_forwarded_for)) {
					local.ipAddress = request.cgi.http_x_forwarded_for;
				}
				if (ListFind(application.wheels.ipExceptions, local.ipAddress)) {
					local.makeException = true;
				}
			}
		}
		if (!local.makeException) {
```

So I can switch my site into maintenance mode... and then anyone in the outside world can just bypass it if they want by providing `?except=[their IP address]` in the URL?

That's... less good than it could be, right?

I get the desirability of being able to bypass maintenance mode for internal / sanctioned users - it is indeed what I was needing to do when I found this code - but this is def not the way to do it.

Suggestions:
* maintain an array of exception IP addresses as a setting _in the codebase_. This way the application owns the exceptions, not the general public.
* do the same password check that is in place for switching modes in the first place.

This is, btw, another example of where the CFWheels lack of OO design/architecture lets it down. In a well-designed application, the `handleMaintenanceMode` or `handleIpExceptions` functionality would be encapsulated in its own class or even just its own function. That way ppl concerned about this flaw could simply override that one method to make it more secure. But the way CFWheels is "designed", I need to override the entire `$runOnRequestStart` method. Ah well: at least it's not the 1000-line `onApplicationStart`... ;-)

I realise CFWheels is unlikely to get any further development so there's little point making remedial suggestions here. I'm rather more posting so ppl using CFWheels know about it.

Cheers.

-- 
Adam

**Comment by MvdO79** [comment]
> I realise CFWheels is unlikely to get any further development.

Is that so?

**Reply by dbelanger** [reply]
Good morning @MvdO79.  No it isn't true, the project is still in development.  Sometimes Adam can be a little abrupt with the project maintainers.

While continuing to be Open Source, the project is now being funded by an [Open Collective project](https://opencollective.com/cfwheels) with Peter Amiri at the helm.  This will of course help a great deal getting bug fixes and new features completed more quickly. He has a developer maintaining version 2 but they're still accepting pull requests of course as it will always be Open Source.

There is also planning going on about version 3 and the best way to approach it.

**Reply by MvdO79** [reply]
Good morning @dbelanger . Ah that's great to hear, looking forward to the further development of Cfwheels. Gonna check the project.

**Reply by adamcameron** [reply]
Well.

Saying it's "still in development" is kinda like saying the patient is still alive after the machine has been switched off because they're still reflexively breathing, I think.

Peter's financial contribution is... definitely a thing. However that balance only benefits the Open Collective Project unless something's being done with it.

My comment was rather more based on reality rather than aspiration.

Look at the pull requests:
https://github.com/cfwheels/cfwheels/pulls

26 open. Some getting towards being a year old.

When was the most recent pull request merged?
https://github.com/cfwheels/cfwheels/pulls?q=is%3Apr+is%3Aclosed

March.

Commits *at all* in the last 12 months?
https://github.com/cfwheels/cfwheels/graphs/contributors?from=2022-08-10&to=2023-08-09&type=c

Not. Much. Going. On.

If one looks back to when Tom King was still involved in the project - 2021:
https://github.com/cfwheels/cfwheels/graphs/contributors?from=2021-01-01&to=2021-12-31&type=c

We see a chunk of work from Tom, but only superficial activity from anyone else. This is pretty common on open source projects; but if the main dev leaves... not promising.

There is no significant work being done on the project. If there was actually any planning going on for v3 - as Dave suggests - I'd expect it to be being discussed here, but this is pretty much a ghost town as well.

One can also look at the CFML zeitgeist. The only horse in town really is ColdBox these days. It's a good, easy to use, thoughtfully-implemented, with a large and engaged community, and most important of all *very actively supported and maintained*. It's one thing carefully maintaining one's existing CFWheels projects, but it would seem a curious decision to *start* a new CFWheels project these days.

I say this with no malice whatsoever despite how Dave might like to position it. I use CFWheels on a daily basis, but otherwise don't have a dog in its fight (nor, btw, do I have any dog in ColdBox's fight), so perhaps my view is a bit less subjective / emotional about things.

**Comment by edthetechie** [comment]
I thought Peter had taken over the project but he's been really quiet
recently.

**Comment by bpamiri** [comment]
Hi everyone, I must apologize for being very silent for the past few months. I thought it was time to resurface and explain myself. Everything David mentioned below is true so why the radio silence you might ask? Well, some people close to me know this already but for the general community let me bring you up to speed.

Late last year, we (my day job) got hit with ransomware. It turned out to be caused by a massive zero day vulnerability in the firewall hardware we were using. This allowed the thread actors to come in and reck havoc. We have spent the better part of the last half a year putting in place systems and procedures that can help mitigate the risk and allow us to recover from future events. Before you ask, yes we had backups, we had snapshots, we had offsite storage of the backups. But when the hardware you rely on to secure your island is itself compromised, it sends you back to the drawing boards.

So with that out of the way, know that there has been development on the framework while I was busy dealing with life. One look at the PRs shows we have accumulated a fairly large number of PRs that will probably take a few point releases to get through. So expect some point releases coming that will deal with some framework optimization as well as some bug fixes.

We have also had some internal conversations about what a CFWheels 3.0 update might look like. There has been considerable work done on some various projects. Integration of TestBox and WireBox have had some work already. As well as restructuring the file hierarchy and being able to move the framework out of the webroot. Some of these are experimental at this stage but work has taken place on them.

Another theoretical exercise we are investigating is the feasibility of rearchitecting CFWheels as plugins for the ColdBox framework. The idea would be to build on the shoulders of others and looking into the feasibility of keeping the developer facing API the same but utilize the ColdBox ecosystem. This is even more experimental at this stage than the restructure work mentioned above.

As for Adam’s comment, I could see how Adam might have gotten that impression. He and I had a conversation last year when he had discovered some errors in the 1.4.x version of the docs and I was able to correct it for him. In that exchange he did ask if there was any value in sending me issues he finds in the 1.4.x branch, you see he is still using that version of CFWheels in his project, and I believe I may have said at the time that I didn’t see us doing much further development on the 1.4.x branch outside of some catastrophic security patch that we may have to release if it became absolutely necessary.

So there you have it. This email is probably way overdue but it accounts for the radio silence during the last several months, the state of the project as of right now, and what we are doing to look towards the future. I appreciate everyone’s patience while I worked through the issues at work and I’m looking forward to getting more activity on the project in the months to come.

-Peter AmiriOn Aug 7, 2023, at 6:36 AM, MvdO79 ***@***.***> wrote:﻿
Good morning @dbelanger . Ah that's great to hear, looking forward to the further development of Cfwheels. Gonna check the project.

**Reply by adamcameron** [reply]
Gonna buy you some `<p>` tags for your birthday, Pete ;-)

Interesting what you say about v3 and planning it "internally". Why's that? In a very small community like this, I'd've thought you'd be wanting as much input and buy-in from what few ppl are around? If nothing else it would probably stave off questions about whether the project is in maintenance mode.

The things you list as developments seem to be the same things I discussed with Tom a year or so ago. Indeed the "rearchitecting CFWheels as plugins for the ColdBox framework" was something *I* suggested to him, instead of redoing the architecting himself (because it *was* just him doing the work).  Is that what you're referring to? The work he did back then? Or is it actively progressing? Have there _really_ been any meaningful "internal planning" conversations since he left?

**Reply by bpamiri** [reply]
Thanks. Replying to the thread on the mail app on the iPad didn't work as well as I expected. All the line breaks were striped out as you saw. Anyway, at my age all take all the birthdays and birthday gifts as I can. So I'll be expecting a gift next February.  ;-)

The discussions I referred to occurred on our private slack channel but you are absolutely right, as an open source project I should try to move those into the public and will try to do so. But historically the project's challenge has been how to engage our small band of followers without ushering in the floodgates of negativism which the CF community seems to have towards the project. 

Frankly I don't get it. At my day job we have a CFWheels based ecommerce B2B site. We process well over 9 figures of revenues annually on code written in CF with the CFWheels framework running on Lucee. I would think something like this would be celebrated by the CF community. It is a testimony that CF is not dead and that CFWheels despite all its architectural issues can be used to generate revenue for a company. 

I also understand your predicament, you are stuck with a 1.4.x codebase. So when you come across a bug or anomaly you start digging into the framework code, in essence into the /wheels directory. I think you are in the 1% in this case, I would venture to guess that most CFWheels developers never dig into the /wheels directory. Anyway, let me continue, while you are digging into the /wheels directory you come across some code that you think could have been developed better. You copy and paste that code into a public forum which welcomes ridicule from a handful of others in the CF community. I've seen this over and over but I've also seen the analysis you do and the amount of work you put into your posts. I would be trilled if someone so passionate would put some of that passionate into submitting PRs.

You say you don't have a horse in this race, I think that is my point. I think you should have a horse in this race, frankly take over the whole racetrack if you want to, because you have so much to contribute and the project would be better off for such effort.

**Reply by timsayshey** [reply]
The news about the Coldbox plugins is exciting! The future of CFWheels is looking bright. It's understandable that such a huge undertaking will take time but it will be totally worth the wait.

**Reply by neokoenig** [reply]
Amazing how life gets in the way of open source projects :)

I've still got wheels running in my main major project, and whilst we're moving to node, that's purely for lack of (CFML) developer resources and a wish to unify the tech stack (we're very heavily VueJS focussed) so that a JS dev can flick between the two. But wheels has allowed me to do (very) rapid development in a start up, and we're still here 5 years later with about 50 people in the company (so we must have been doing something right).

Interestingly my work with [NestJS](https://nestjs.com/) has given me a load of ideas of what you could do with wheels. It's much more angular based though with decorators etc. I think moving languages / frameworks from time to time is really healthy and gives you loads more insight into things.

I've never used coldbox, but I can't imagine it's going to be "that" hard to essentially create syntactic sugar infront of some of the core modules like routing etc. I can imagine the ability to use `findAll()` as an alias to something like quick ORM, (where you could still drill down the core the quick ORM object if wanted / needed) would be a nice addition. The routing translation should be relatively trivial.

In my mind the biggest decisions to be made revolve around to what extent you should change the current syntax/lack of scoping. i.e, should all wheels functions be prefixed with `wheels.` (or `w.` or whatever); and should you enforce any variables which are to be used in view files to be be in the `req` scope (or whatever coldbox uses). there are some very good reasons for doing both of these, not least as it would then play more nicely with coldbox - but would allow controller action reuse, easier testing and cleaner/clearer controllers. Naturally you kill backwards compatibility, but then I think wheels 3 would be the correct time to do it.

**Comment by smilebringergames** [comment]
I personally love cfwheels in the older 1.4.x form because it's not as
difficult to get a page set up (newer version is difficult with routing)
for stuff. So at home I run with that.

It's a shame it's not getting more updates. Maybe I should put some time in
understanding the code so I can update it.

As for the popularity of ColdFusion, I think a lot of people are just on
ASP.net and PHP etc.

The company I work at uses a combination of Adobe ColdFusion, Lucee and as
framework an in-house developed system with Bootstrap on top.

I know companies that still use it, but of course it's not as popular as
the bigger one's. On my YouTube channel (programming tutorials) I get
mostly requests and questions from students at universities who challenge
themselves with learning different web languages. They even asked me to
tutor them. Haven't had time for such however.

Just my feedback to the last few emails on this thread.

Steven
@smilebringergames
